### PR TITLE
Mock LayerMetadataRestorer in analzyer unit tests

### PR DESCRIPTION
**Pull request by jabrown85**
- Moved tests to LayerMetadataRestorer that were not already covered
- Introduced mock and expectations to remaining tests

Signed-off-by: Jesse Brown <jabrown85@gmail.com>

Address #587 

I started in on removing the concrete metadata restorer in the restorer unit tests but stopped. Those test rely heavily on the files being present to ensure that files were deleted or left alone. I felt like moving to the mock would mean that all the code to setup those tests would rely on the tests creating static files on the fly that would be more brittle. Let me know what you think.

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -157,6 +157,7 @@ github.com/google/go-cmp v0.5.5 h1:Khx7svrCpmxxtHBq5j2mp/xVjsi8hQMfNLvJFAlrGgU=
 github.com/google/go-cmp v0.5.5/go.mod h1:v8dTdLbMG2kIc/vJvl+f65V22dbkXbowE6jgT/gNBxE=
 github.com/google/go-containerregistry v0.4.1 h1:Lrcj2AOoZ7WKawsoKAh2O0dH0tBqMW2lTEmozmK4Z3k=
 github.com/google/go-containerregistry v0.4.1/go.mod h1:Ct15B4yir3PLOP5jsy0GNeYVaIZs/MK/Jz5any1wFW0=
+github.com/google/go-containerregistry v0.5.0 h1:eb9sinv4PKm0AUwQGov0mvIdA4pyBGjRofxN4tWnMwM=
```
Comment (jabrown85) [inline go.sum:160]: Something must be off in `main` - this showed up in a couple of PRs today 🤔

File: analyzer_test.go [hunk 2]
```diff
@@ -118,532 +114,56 @@ func testAnalyzerBuilder(platformAPI string) func(t *testing.T, when spec.G, it
...
-
-						_, err := analyzer.Analyze()
-						h.AssertNil(t, err)
-
-						h.AssertPathDoesNotExist(t, filepath.Join(layersDir, "metadata.buildpack"))
-					})
-				})
-
 				it("returns the analyzed metadata", func() {
+					metadataRestorerExpectations(metadataRestorer, analyzer, expectedAppMetadata, expectedCacheMetadata)
```
Comment (jabrown85) [inline analyzer_test.go:118]: I really didn't like having to do this in multiple places, but I couldn't find a clean way to either clear expectations when I wanted to set a new one or to defer setting up the expectation until right before each test with the right context. Open to ideas though!

File: tools/go.sum [hunk 3]
```diff
@@ -246,6 +246,8 @@ github.com/google/go-cmp v0.5.5 h1:Khx7svrCpmxxtHBq5j2mp/xVjsi8hQMfNLvJFAlrGgU=
 github.com/google/go-cmp v0.5.5/go.mod h1:v8dTdLbMG2kIc/vJvl+f65V22dbkXbowE6jgT/gNBxE=
 github.com/google/go-containerregistry v0.4.1 h1:Lrcj2AOoZ7WKawsoKAh2O0dH0tBqMW2lTEmozmK4Z3k=
 github.com/google/go-containerregistry v0.4.1/go.mod h1:Ct15B4yir3PLOP5jsy0GNeYVaIZs/MK/Jz5any1wFW0=
+github.com/google/go-containerregistry v0.5.0 h1:eb9sinv4PKm0AUwQGov0mvIdA4pyBGjRofxN4tWnMwM=
+github.com/google/go-containerregistry v0.5.0/go.mod h1:Ct15B4yir3PLOP5jsy0GNeYVaIZs/MK/Jz5any1wFW0=
```
Comment (yaelharel) [inline tools/go.sum:250]: Related to https://github.com/buildpacks/lifecycle/pull/605/files#r626117500.

File: analyzer_test.go [hunk 4]
```diff
@@ -102,10 +97,17 @@ func testAnalyzerBuilder(platformAPI string) func(t *testing.T, when spec.G, it
...
 		when("#Analyze", func() {
 			var (
-				expectedAppMetadata platform.LayersMetadata
-				ref                 *testmock.MockReference
+				expectedAppMetadata   platform.LayersMetadata
+				expectedCacheMetadata platform.CacheMetadata
+				ref                   *testmock.MockReference
 			)
 
+			metadataRestorerExpectations := func() {
```
Comment (natalieparellano) [inline analyzer_test.go:105]:
```suggestion
			expectRestoresMetadata := func() {
```

**Comment by natalieparellano** [comment]
> I started in on removing the concrete metadata restorer in the restorer unit tests but stopped. Those test rely heavily on the files being present to ensure that files were deleted or left alone. I felt like moving to the mock would mean that all the code to setup those tests would rely on the tests creating static files on the fly that would be more brittle. Let me know what you think.

I think this makes sense. We might consider removing or condensing the following assertions in restorer_test.go as they are all duplicated (I **think**) in layer_metadata_restorer_test.go: https://github.com/buildpacks/lifecycle/blob/06bc7289a984b54e65403614f171cc844f361ea8/restorer_test.go#L500-L588

File: layer_metadata_restorer_test.go [hunk 5]
```diff
@@ -150,6 +175,24 @@ func testLayerMetadataRestorer(t *testing.T, when spec.G, it spec.S) {
 				}
 			})
 
+			when("cache with inconsistent metadata exists", func() { // cache was manipulated or deleted
```
[the remark this answers is not in the document]
Comment (natalieparellano) [reply layer_metadata_restorer_test.go:178]: Could we move this to the bottom of the parent `when`? (for readability)

**Comment by natalieparellano** [comment]
It still kind of bothers me that restorer_test.go has to care about both `buildpackAPI` and `platformAPI`. But I'm not sure if there is a way to fix it without pulling the "cleaning" of the layer and sha files into a separate unit-tested service. We could put that in a separate issue if we think it's a good idea.

File: analyzer_test.go (hunk 4 above)
Comment (jabrown85) [reply analyzer_test.go]: a432198

File: layer_metadata_restorer_test.go [hunk 6]
```diff
@@ -127,6 +133,25 @@ func testLayerMetadataRestorer(t *testing.T, when spec.G, it spec.S) {
...
 					// Cache-image-only layers.
 					{"metadata.buildpack/cache.toml", "[metadata]\n  cache-key = \"cache-value\""},
+				} {
+					got := h.MustReadFile(t, filepath.Join(layerDir, data.name))
+					h.AssertStringContains(t, string(got), data.want)
+					h.AssertStringDoesNotContain(t, string(got), unsetFlags) // The [types] table shouldn't exist. The build, cache and launch flags are set to false.
+				}
+			})
+
+			it("restores layer metadata and preserves the values of the launch, build and cache flags in top level", func() {
```
[the remark this answers is not in the document]
Comment (jabrown85) [reply layer_metadata_restorer_test.go]: Fixed in a432198. I decided not to wrap in `>= 0.6` but if you feel strongly I can.

File: layer_metadata_restorer_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jabrown85) [reply layer_metadata_restorer_test.go]: a432198

**Comment by jabrown85** [comment]
I agree @natalieparellano about the `restorer_test.go` having to spin through both API sets. I think I would like to see us eventually pull most of the branch logic into smaller units of work so we can simply assert it was called with the given version from `restorer_test.go`. I think we can incrementally do that work.

File: restorer_test.go [hunk 7]
```diff
@@ -496,95 +496,6 @@ func testRestorerBuilder(buildpackAPI, platformAPI string) func(t *testing.T, wh
 					metadata := h.MustReadFile(t, filepath.Join("testdata", "analyzer", "app_metadata.json"))
 					h.AssertNil(t, json.Unmarshal(metadata, &restorer.LayersMetadata))
 				})
-
```
Comment (yaelharel) [inline restorer_test.go:499]: Can you please explain why it's ok to delete the tests below?
Maybe I missed it but I didn't see that you moved them to a different file.

Comment (jabrown85) [reply restorer_test.go:499]: See https://github.com/buildpacks/lifecycle/pull/605#issuecomment-838923343 - but basically these were already brought over in the layer metadata analyzer tests when it was written, but left here initially as well.

File: analyzer_test.go [hunk 8]
```diff
@@ -118,532 +120,55 @@ func testAnalyzerBuilder(platformAPI string) func(t *testing.T, when spec.G, it
...
-						{"metadata.buildpack/launch.sha", ""},
-						{"metadata.buildpack/launch-build-cache.sha", ""},
-						{"metadata.buildpack/launch-cache.sha", ""},
-						{"no.cache.buildpack/some-layer.sha", ""},
-					} {
-						h.AssertPathDoesNotExist(t, data.name)
-					}
-				})
-
-				it("does not restore launch=false layer metadata", func() {
```
Comment (yaelharel) [inline analyzer_test.go:193]: Please see https://github.com/buildpacks/lifecycle/pull/605/files#r631126338

File: restorer_test.go (hunk 7 above)
Comment (yaelharel) [reply restorer_test.go:499]: ohh, ok. I'll take a look. Thanks!

File: layer_metadata_restorer_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (natalieparellano) [reply layer_metadata_restorer_test.go]: I think wrapping would be more readable... or at least, we should be consistent.

**Review by natalieparellano — APPROVED** [review]
Had one nit about wrapping a context in layer_metadata_restorer_test.go. It's great to see where this ended up!

**Comment by natalieparellano** [comment]
@jabrown85 do you mind fixing the merge conflicts?

File: layer_metadata_restorer_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (jabrown85) [reply layer_metadata_restorer_test.go]: I added the `without the launch...` test in 8ac59fa

File: layer_metadata_restorer_test.go [hunk 9]
```diff
@@ -130,9 +136,52 @@ func testLayerMetadataRestorer(t *testing.T, when spec.G, it spec.S) {
...
+						h.AssertStringContains(t, string(got), data.want)
+					}
+				})
+			})
+
+			when("buildpack api >= 0.6", func() {
+				it("restores layer metadata without the launch, build and cache flags", func() {
+					buildpacks = []buildpack.GroupBuildpack{
+						{ID: "metadata.buildpack", API: "0.6"},
+						{ID: "no.cache.buildpack", API: "0.6"},
```
Comment (yaelharel) [inline layer_metadata_restorer_test.go:167]: Maybe use the latest buildpack API instead of `0.6`?

File: analyzer_test.go [hunk 10]
```diff
@@ -118,532 +120,55 @@ func testAnalyzerBuilder(platformAPI string) func(t *testing.T, when spec.G, it
...
-
-					when("buildpack API < 0.6", func() {
-						it.Before(func() {
-							analyzer.Buildpacks = []buildpack.GroupBuildpack{
-								{ID: "metadata.buildpack", API: "0.5"},
-								{ID: "no.cache.buildpack", API: "0.5"},
-							}
-						})
-
-						it("restores app and cache layer metadata and preserves the values of the launch, build and cache flags", func() {
```
Comment (yaelharel) [inline analyzer_test.go:432]: Do we continue testing it?
I couldn't find this test anywhere.

File: analyzer_test.go [hunk 11]
```diff
@@ -118,532 +120,55 @@ func testAnalyzerBuilder(platformAPI string) func(t *testing.T, when spec.G, it
...
-					it("does not restore cache=false layer metadata", func() {
-						h.SkipIf(t, api.MustParse(platformAPI).Compare(api.MustParse("0.7")) >= 0, "Platform API >= 0.7 does not restore layer metadata")
-
-						_, err := analyzer.Analyze()
-						h.AssertNil(t, err)
-
-						h.AssertPathDoesNotExist(t, filepath.Join(layersDir, "metadata.buildpack", "cache-false.toml"))
-					})
-
-					it("returns a nil image in the analyzed metadata", func() {
```
Comment (yaelharel) [inline analyzer_test.go:577]: Same as https://github.com/buildpacks/lifecycle/pull/605/files#r634546047.

File: analyzer_test.go (hunk 10 above)
Comment (jabrown85) [reply analyzer_test.go:432]: https://github.com/buildpacks/lifecycle/pull/605/files#diff-6834a3c40b637f9d72a33e93feaa8db461b3551f15e6a1fffa1a5c6dd6cd4ccdR144

File: analyzer_test.go (hunk 11 above)
Comment (jabrown85) [reply analyzer_test.go:577]: https://github.com/buildpacks/lifecycle/pull/605/files#diff-67271f60deaa2e51dcee16995bc6eb5e599dc120dd47dddc41137dca5a1a6509R159

File: analyzer_test.go (hunk 10 above)
Comment (yaelharel) [reply analyzer_test.go:432]: You're right!
I search for `launch=true` instead of `launch = true`.
Thanks!

File: analyzer_test.go (hunk 11 above)
Comment (yaelharel) [reply analyzer_test.go:577]: Again, thanks!

**Review by yaelharel — APPROVED** [review]

**Pull request merged**